File: 937 - Reorder Data in Log Files.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cctype>

using namespace std;

bool compare_log(const string & a, const string & b) {
    string as[2], bs[2];
    as[0] = a.substr(0, a.find(' '));
    as[1] = a.substr(a.find(' '));
    bs[0] = b.substr(0, b.find(' '));
    bs[1] = b.substr(b.find(' '));
    bool ad = isdigit(as[1][1]), 
        bd = isdigit(bs[1][1]);
    if (!ad && !bd) {
        if (as[1] != bs[1]) {
            return as[1] < bs[1];
        } else {
            return as[0] < bs[0];
        }
    } else {
        return ad ? false : true;
    }
}

class Solution {
public:
    vector<string> reorderLogFiles(vector<string>& logs) {
        stable_sort(logs.begin(), logs.end(), compare_log);
        return logs;
    }
};

File: 937 - Reorder Data in Log Files.java
import java.util.Arrays;

class Solution {
    public String[] reorderLogFiles(String[] logs) {
        Arrays.sort(logs, (a, b) -> {
            String[] as = a.split(" ", 2), bs = b.split(" ", 2);
            boolean ad = Character.isDigit(as[1].charAt(0)), 
                    bd = Character.isDigit(bs[1].charAt(0));
            if (!ad && !bd) {
                int cmp = as[1].compareTo(bs[1]);
                if (cmp == 0) {
                    return as[0].compareTo(bs[0]);
                } else {
                    return cmp;
                }
            } else {
                return ad ? (bd ? 0 : 1) : -1;
            }
            
        });
        return logs;
    }
}